### ISSUE TITLE
Support SwiftUI's spring animation

### ISSUE THREAD

**Issue by dkwingsmt** [flutter/flutter#152587]
### Use case

The spring animation is used all over iOS, and is critical to building widgets with iOS look and feel.

### Proposal

The spring animation is introduced here: https://developer.apple.com/videos/play/wwdc2023/10158/

**Comment by davidhicks980** [comment]
I'm excited to see this on the radar. Assuming this would be a subclass of Curve, one thing I'm wondering is how the animation duration specified by the AnimationController would behave, considering Curves don't have access to the duration of their driving AnimationController?

```dart
animationController.animateTo(
  1,
  duration: const Duration(milliseconds: 500),
  curve: SpringCurve(bounce: 0.2, duration: 0.5), // Considering this is perceptual time, the spring wouldn't settle prior to stopping
)
```

**Comment by jodierizky** [comment]
I've been waiting a long time for this

**Comment by dkwingsmt** [comment]
Update: It turns out that Flutter has long supported spring simulations, but simulations are not integrated into other components well. https://github.com/flutter/flutter/pull/155575 will allow `AnimationController` and `TransitionRoute` to use simulations.

Interestingly, this PR won't help much on the fidelity of Cupertino dialog or action sheets, because the current curve is close enough. However, using the proper simulation is still useful because it allows non-zero initial velocity, such as when the animation switches half way. Although, I don't know many specific cases for this.

After https://github.com/flutter/flutter/pull/155575 there will still be some work left. One of them is some convenient APIs that constructs `SpringDescription` in ways that replicate the SwiftUI APIs. I've done some research, and if I remember correctly, you can simply use the numbers extracted from XCode if you just want to replicate some native behavior. 

What I'm trying to say is, I'm not sure what is badly needed for this issue after https://github.com/flutter/flutter/pull/155575. So let me know if there are any specific issues that are bothering you. Thanks!

**Comment by HaijunWei** [comment]
I have a question. 
Using the above formula, the result is wrong.
``` Swift
duration = 0.5，bounce = 0.2
// Wrong formula
damping = 1 - 4 * pi * bounce / duration        // -4.026548245743669

// The right formula, The correct value is obtained by the method described in Apple's documentation
// https://developer.apple.com/documentation/swiftui/spring
damping = 4 * pi * (1 - bounce) / duration        // 20.106192982974676
```

bounce < 0, I don't know how to calculate the correct value.

**Comment by dkwingsmt** [comment]
I've found this as well, and I've wasted so much time on it. The formulas were screenshot from the video and I don't know why they said so.

My suggestion is to extract basic properties (mass, stiffness, damping) from a built spring animation object and use them in Flutter.

**Comment by HaijunWei** [comment]
I agree, Apple seems to be misleading developers.
I found that the algorithm of this library is closer to the curve of SwiftUI.

https://github.com/sodality-tech/figma_spring_curve/blob/main/lib/src/figma_spring_curve.dart

**Comment by orestesgaolin** [comment]
Regarding this part of the @dkwingsmt comment:

> such as when the animation switches half way. Although, I don't know many specific cases for this.

Perhaps not Flutter built-in solution, but I found that you can use custom animation controller to drive the bottom sheet.

I used [flutter_physics](https://github.com/sangddn/flutter_physics) but probably any animation controller following the spring simulation should work similarly.

Before

https://github.com/user-attachments/assets/21586a85-c2a8-4dec-a074-5f854132c374

After



https://github.com/user-attachments/assets/74ced382-0fad-419b-801a-909d2da662b0


Sharing the adapter here as it may be useful to other people.


<details><summary>Code sample</summary>

```dart
import 'package:flutter/material.dart';
import 'package:flutter_physics/flutter_physics.dart';

class AnimationsControllerAdapter implements AnimationController {
  AnimationsControllerAdapter(this.physicsController);

  final PhysicsController physicsController;

  @override
  Duration? get duration => physicsController.duration;

  @override
  set duration(Duration? value) {
    physicsController.duration = value;
  }

  @override
  Duration? get reverseDuration => physicsController.reverseDuration;

  @override
  set reverseDuration(Duration? value) {
    physicsController.reverseDuration = value;
  }

  @override
  double get value => physicsController.value;

  @override
  set value(double value) {
    physicsController.value = value;
  }

  @override
  void addListener(VoidCallback listener) {
    physicsController.addListener(listener);
  }

  @override
  void addStatusListener(AnimationStatusListener listener) {
    physicsController.addStatusListener(listener);
  }

  @override
  TickerFuture animateBack(double target,
      {Duration? duration, Curve curve = Curves.linear}) {
    return physicsController.animateTo(
      target,
      duration: duration,
      physics: curve,
    );
  }

  @override
  TickerFuture animateTo(double target,
      {Duration? duration, Curve curve = Curves.linear}) {
    return physicsController.animateTo(
      target,
      duration: duration,
      physics: curve,
    );
  }

  @override
  TickerFuture animateWith(Simulation simulation) {
    return physicsController.animateWith(simulation);
  }

  @override
  AnimationBehavior get animationBehavior =>
      physicsController.animationBehavior;

  @override
  void clearListeners() {
    physicsController.clearListeners();
  }

  @override
  void clearStatusListeners() {
    physicsController.clearStatusListeners();
  }

  @override
  String? get debugLabel => physicsController.debugLabel;

  @override
  void didRegisterListener() {
    // TODO: implement didRegisterListener
  }

  @override
  void didUnregisterListener() {
    physicsController.didUnregisterListener();
  }

  @override
  void dispose() {
    physicsController.dispose();
  }

  @override
  Animation<U> drive<U>(Animatable<U> child) {
    return physicsController.drive(child);
  }

  @override
  TickerFuture fling(
      {double velocity = 1.0,
      SpringDescription? springDescription,
      AnimationBehavior? animationBehavior}) {
    return physicsController.fling(
      velocity: velocity,
      springDescription: springDescription,
      animationBehavior: animationBehavior,
    );
  }

  @override
  TickerFuture forward({double? from}) {
    return physicsController.forward(from: from);
  }

  @override
  bool get isAnimating => physicsController.isAnimating;

  @override
  bool get isCompleted => physicsController.isCompleted;

  @override
  bool get isDismissed => physicsController.isDismissed;

  @override
  bool get isForwardOrCompleted => physicsController.isForwardOrCompleted;

  @override
  Duration? get lastElapsedDuration => physicsController.lastElapsedDuration;

  @override
  double get lowerBound => physicsController.lowerBound;

  @override
  void notifyListeners() {
    physicsController.notifyListeners();
  }

  @override
  void notifyStatusListeners(AnimationStatus status) {
    physicsController.notifyStatusListeners(status);
  }

  @override
  void removeListener(VoidCallback listener) {
    physicsController.removeListener(listener);
  }

  @override
  void removeStatusListener(AnimationStatusListener listener) {
    physicsController.removeStatusListener(listener);
  }

  @override
  TickerFuture repeat(
      {double? min,
      double? max,
      bool reverse = false,
      Duration? period,
      int? count}) {
    return physicsController.repeat(
      min: min,
      max: max,
      reverse: reverse,
      period: period,
      count: count,
    );
  }

  @override
  void reset() {
    physicsController.reset();
  }

  @override
  void resync(TickerProvider vsync) {
    physicsController.resync(vsync);
  }

  @override
  TickerFuture reverse({double? from}) {
    return physicsController.reverse(from: from);
  }

  @override
  AnimationStatus get status => physicsController.status;

  @override
  void stop({bool canceled = true}) {
    physicsController.stop(canceled: canceled);
  }

  @override
  String toStringDetails() {
    return physicsController.toStringDetails();
  }

  @override
  TickerFuture toggle({double? from}) {
    throw UnimplementedError('toggle is not implemented yet');
  }

  @override
  double get upperBound => physicsController.upperBound;

  @override
  double get velocity => physicsController.velocity;

  @override
  Animation<double> get view => physicsController.view;
}

class BottomSheetPage extends StatefulWidget {
  const BottomSheetPage({super.key});

  @override
  State<BottomSheetPage> createState() => _BottomSheetPageState();
}

class _BottomSheetPageState extends State<BottomSheetPage>
    with TickerProviderStateMixin {
  late PhysicsController controller;

  @override
  void initState() {
    controller = PhysicsController(vsync: this);
    super.initState();
  }

  @override
  void dispose() {
    controller.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: Center(
        child: ElevatedButton(
          onPressed: () {
            showModalBottomSheet(
              context: context,
              transitionAnimationController:
                  AnimationsControllerAdapter(controller),
              builder: (context) {
                return Container(
                  height: 200,
                  color: Colors.blueGrey,
                  child: Center(
                    child: Column(
                      mainAxisAlignment: MainAxisAlignment.center,
                      mainAxisSize: MainAxisSize.min,
                      children: <Widget>[
                        const Text('Modal BottomSheet'),
                        ElevatedButton(
                          child: const Text('Close BottomSheet'),
                          onPressed: () => Navigator.pop(context),
                        ),
                      ],
                    ),
                  ),
                );
              },
            );
          },
          child: Text('Show bottom sheet'),
        ),
      ),
    );
  }
}
```

</details>

**Comment by bernaferrari** [comment]
I think Flutter is missing `duration` and `bouce` in SpringSimulation so it is similar to Swift. Could be even SpringSimulation.timed({double duration, double bounce}) so that user doesn't need to know what stiffness is.

**Comment by dkwingsmt** [comment]
@bernaferrari That's awesome! If it's useful to you we can definitely add it. The `withDurationAndBounce` name is ok, or if you can think of something shorter :)  I'd still like to figure out the formula for negative bounces though, but it's ok if we leave it as a TODO.

**Comment by bernaferrari** [comment]
Now the only thing preventing me from being truly happy is `Curve.spring()` or something like that so I can use it everywhere 👀

Some curves are really simple. Could just adapt it to accept mass/stiffness/damping or duration/bounce, or all of them but asserting:
```dart
class _BounceInCurve extends Curve {
  const _BounceInCurve._();

  @override
  double transformInternal(double t) {
    return 1.0 - _bounce(1.0 - t);
  }
}
```

**Comment by dkwingsmt** [comment]
It's not going to work as a curve, because curve has a fixed end while the spring doesn't. First, a spring never actually stops, but only settles to within a tolerance. Second, even for the same spring the curve varies with starting velocity.

However `AnimationController` already supports simulation. Can you investigate what's preventing you to use the simulation just as happily?

**Comment by bernaferrari** [comment]
I'm using PageView's nextPage (https://github.com/flutter/flutter/blob/81139dbcab9a7087a0febf027fd183509f8ae628/packages/flutter/lib/src/widgets/page_view.dart#L246) which only supports a Curve, so I synced up its time with `withDurationAndBounce` duration and it worked fine. As long as the bounce is not very large it seems to settle correctly/fine.

I also got an `AnimatedContainer` I was hoping to apply the spring, so for example color changes are alive like Swift does.

**Comment by dkwingsmt** [comment]
We should make sure that everywhere that supports `curve` should also support simulation. For example, this API should probably allow either `curve` or `simulation` as optional parameters.

**Comment by bernaferrari** [comment]
Are you sure? There is a massive amount of places that use curve without simulation (plus every library out there).

For example, `AnimatedSwitcher` has `switchInCurve` and `switchOutCurve`, would it have `switchInSimulation` `switchOutSimulation`? Then there is `AnimatedCrossFade` with 3 curves, `AnimatedOpacity`, `RenderAnimatedSize`, `TabController.animateTo`. There might be like 20-30 usages in Flutter (totally guessing).

I'm totally in favor of supporting springs, just double checking if that's the easiest/best way.

**Comment by dkwingsmt** [comment]
I think the biggest different between a curve animation and a simulation animation is that a simulation doesn't have a predefined end, or even a fixed curve at all. Even though we know that a default spring simulation is created with a `duration: 500ms` and its settling duration is 414ms _with a zero starting velocity_, the animation can vary greatly depending on the releasing velocity, such as the user's gestures, which is a runtime information. For example, imagine when the action sheet is closed at zero starting velocity vs when the user fiercely shuts it down. The animation must start with the gesture's swiping velocity.

If we know that a widget's animation will always start at 0 velocity, such as the dialogs' opening and closing, then we can approximate it with a curve. In fact, Flutter has the curve `linearToEaseOut`, which is a really good approximation even before https://github.com/flutter/flutter/pull/155575. For such curves, we can certainly provide a utility class to convert `SpringDescription` to a curve.

**Comment by dkwingsmt** [comment]
I just realized that there's an `AnimationStyle` class that encapsulates animation related properties. If we update widgets to `AnimationStyle` (which we should eventually) then supporting simulation is a piece of cake!

**Comment by bernaferrari** [comment]
I was experimenting with adding SpringAnimationController so it worked with fling and the problematic cases. Still, even if I successfully gotten SpringCurve to work, there is a high chance we would need if curve is Spring or something like that to determine the simulation, and this would be against flutter guidelines. Unless we have SpringAnimationController and AnimationController, but then it would be weird/hard to expose them where they need to go.

**Comment by dkwingsmt** [comment]
What do you mean by `SpringCurve`? Does it only handle springs with 0 start velocity? Or are you somehow refactoring the `Curve` class to accept dynamic curves?

Also why would you want an `AnimationController`? It already supports simulations with `animateWith` and `animateBackWith`.

(Also, I have tagged you on Discord for another issue. It's in the hackers channel, one of the latest threads. Can you take a look?) (Will delete this line after you do.)

**Comment by timcreatedit** [comment]
> I just realized that there's an `AnimationStyle` class that encapsulates animation related properties. If we update widgets to `AnimationStyle` (which we should eventually) then supporting simulation is a piece of cake!

This is also the solution that I would've had in mind for package:springster!

I think overall it is the most explicit solution for users.

**Comment by dkwingsmt** [comment]
My first understanding was that simulation based animation should be the general form, and curves are just a special case of simulations.

| Simulations | Curves |
|----|----|
| Has a fixed start | Has a fixed start |
| End decided at runtime | End decided beforehand |
| Curve decided at runtime | Curve decided beforehand |

The one (yet important) miss of simulations are that they're theoretically unidirectional. I wonder how much impact that _really_ has, because although many widgets need to play animations forward and backward, they can easily be done by two separate simulations instead of really reversible curves. The only case I think curves are irreplaceable is random dropping. For example, a [navigation bar](https://github.com/flutter/flutter/pull/164956) can be dragged left and right and left, requiring the animation controller to be able to know the curve result at every time point.

As a conclusion, I propose the following model: one optional simulation followed by optional curves. That is, the animation controller is always base on either a simulation or a straight line over a fixed duration. On top of that there can be arbitrary numbers of curves to convert the result.

> position should be animated using a 2D spring simulation in this case

This was exactly what was demonstrated in Apple's talk. Flutter should support it eventually.

**Comment by bernaferrari** [comment]
On `Breaking changes that would improve the overall API` list, one of the items is

> Instead of one AnimationController, have one for each type of controller (bounded, unbounded, simulation, repeating, etc).

Maybe this could be the start of that?

`BoundedAnimationController.withSpring()` or who knows, but could make sense.

**Comment by davidhicks980** [comment]
Doesn't the default AnimationController already put bounds on the animation output? Otherwise, ClampedSimulation can wrap a simulation to keep the movement between 0 and 1.

**Comment by bernaferrari** [comment]
Yes, it does, but they regret and wish they could uncouple it. Maybe with spring and these variations, could make sense to finally do this change.

**Comment by timcreatedit** [comment]
> Yes, it does, but they regret and wish they could uncouple it. Maybe with spring and these variations, could make sense to finally do this change.

I think one strong argument for this, is that things like `forward` and `reverse` can't work at all on an unbounded controller that's driven by a simulation, since the simulation needs the end point.

**Comment by bernaferrari** [comment]
But the infinite case is an exception, not the rule. We could just clamp to a fixed time or warn when the settle time is infinite. Most people just want 300, 500ms spring motion. Most people want a button that you can click and then lift using spring.

**Comment by timcreatedit** [comment]
> But the infinite case is an exception, not the rule. We could just clamp to a fixed time or warn when the settle time is infinite. Most people just want 300, 500ms spring motion. Most people want a button that you can click and then lift using spring.

I don't think the unbounded case being an exception is relevant to this architectural decision. You were suggesting to uncouple bounded and unbounded controllers as part of building springs into Flutter, I was only supporting that suggestion with a key benefit that would bring (and a footgun that adding simulations without it would entail)

Also, in principle I agree that animations are mostly from a `lowerBound` to an `upperBound` and back.
However, both in my [spring-based hero package](https://pub.dev/packages/heroine) and [animated_to](https://pub.dev/packages/animated_to) by @chooyan-eng (both use [springster](https://pub.dev/packages/springster) internally), unbounded controllers are used, because animating Widgets from A to B dynamically doesn't make sense in a bounded way.

I am currently working on an experimental rewrite of [springster](https://pub.dev/packages/springster) that aims to unify an API for Duration/Curve, as well as Physics-based simulations. I have also split the types for bounded and unbounded controllers there. I'll publish a pre-release later today, I am quite happy with how it turned out, it might be helpful in the discussion.

**Comment by dkwingsmt** [comment]
We're confusing a couple of terms. The concept of being "unbounded" is being used for different things and we must clear it up.

### "unbounded" as in output

> Doesn't the default AnimationController already put bounds on the animation output? Otherwise, ClampedSimulation can wrap a simulation to keep the movement between 0 and 1.

> in principle I agree that animations are mostly from a lowerBound to an upperBound and back.

These sentences are talking about is changing the _output_, which is trivial. Limiting the output doesn't need an `AnimationController` at all. We just need a `AnimationTransformer<double, double>` that transforms a double in a clamping way.

### "unbounded" as in input time

> I think one strong argument for this, is that things like forward and reverse can't work at all on an unbounded controller that's driven by a simulation, since the simulation needs the end point.

This sentence is talking about the animation persisting even as the time goes to infinity. 

## Is simulation different because it might extend to infinite time?

The answer is no. Let me explain bit by bit.

First of all, sure, 99.9% animations ends (as of a tolerance) after a while. But should you reverse a simulation just because it has an end?

No. You should **never** reverse a simulation, no matter if they have an end.

Imagine we have a critically damped spring, so that you know that the vibration will end in around 500ms, does it mean you _can_ play it back from 500ms back to 0ms? 

Yes you _can_, since we have analytical formula to compute the `x` at any point of time. But _should_ you? If you pull the spring out and let it bounce, it bounces weaker and weaker, eventually stabilizes. Reversing this curve means that that the spring leaves the stable point and bounce stronger and stronger. What does it even mean? We don't allow simulation to be reversed not because mathematical possibility and programmatical possibility. **We don't allow simulation to be reversed because they're banned. They're evil.**

But wait, actually, we should just ban all reversed animations, including curved ones. Think about it: **reversing curves is never needed**. If you have a widget that flies in in some curve, to fly it out you just play another curve with swapped endpoints. The API was there just for convenience, not because it was a good model. In fact, the current API of reversing an animation is actually counterintuitive. Widgets tend to writing
```dart
        CurvedAnimation(
          parent: widget.primaryRouteAnimation,
          curve: Curves.linearToEaseOut,
          reverseCurve: Curves.linearToEaseOut.flipped,
        )
```
than using the same forward curve for `reverseCurve`. What the widgets actually want to do is in fact saying:
* This widget has two phases: fly in and fly out.
* For fly in, play a curve from out to in.
* For fly out, play a curve from in to out.

Do you notice it? Reversibility is never a problem. I could have replaced every word "curve" with "simulation" and it would have work perfectly all the same. **Reversibility is not a deciding difference between curves and simulations.**

## Then what's actually different between `Curve` and `Simulation`?

I've explained it in my previous [comment](https://github.com/flutter/flutter/issues/152587#issuecomment-2731252466), but I might have focused too much on having a definite end. Let me rephrase it: The biggest different between a `Curve` and a `Simulation` is that the result of a `Simulation` relies on initial conditions, i.e. initial velocity. 

The following figure shows how a spring bounces with a zero initial velocity:
<img width="683" alt="Image" src="https://github.com/user-attachments/assets/d023004c-98f5-41c1-b3a3-5b9eef7472f9" />

The following figure shows how a spring bounces with a negative initial velocity:
<img width="706" alt="Image" src="https://github.com/user-attachments/assets/7f37ccdb-14b8-400d-b319-dce2cd78c644" />

They're described by the same `SpringDescription`. But they're different curves. They can't be represented by a fixed `Curve` class. The `Curve` class is limited. Simulations are irreplaceable. In fact, the curve represented by `Curve` classes are special cases of simulations where initial velocity doesn't matter.
